rdrf/rdrf/forms/components.py

```diff
@@ -282,23 +283,54 @@ def _form(context_form_group):
 
         return form_links
 
-    def _get_existing_links(self, context_form_group):
-        links = []
-
-        def is_current(url):
-            parts = url.split("/")
-            context_id = int(parts[-1])
-            if self.current_rdrf_context_model:
-                return context_id == self.current_rdrf_context_model.pk
-            else:
-                return False
-
-        for url, text in self.patient_model.get_forms_by_group(context_form_group):
-            if not text:
-                text = "Not set"
-            link_obj = Link(url, text, is_current(url))
+    def _get_form_list_link(self, form):
+        return reverse("registry_form_list", args=[
+            self.registry_model.code,
+            form.id,
+            self.patient_model.pk,
+        ])
+
+    def _get_existing_links(self, context_form_group, slice_len=5):
+        """
+        Create a subset of context form links of slice_len length, where the
+        currently-selected context form (if one is selected) is as central as
+        possible:
+
+          01234.6789 (slice_len=5)
+        = 34.67
+
+          01234.6 (slice_len=4)
+        = 34.6
+
+          0123456 (slice_len=3)
+        = 012
+
+        :return: links, current_index, total_forms
+        """
+        links = deque(maxlen=slice_len)
+        current_index = -1
+        index_found = False
+        current_context_id = self.current_rdrf_context_model.pk if self.current_rdrf_context_model else None
```

Comment:
There's a bug where selecting a form in another CFG will impact how the other CFGs are listed on the page.